### PR TITLE
[AArch64][Driver] Better handling of target feature dependencies

### PR DESCRIPTION
**Pull request by ostannard**
Currently there are several bits of code in the AArch64 driver which attempt to enforce dependencies between optional features in the -march= and -mcpu= options. However, these are based on the list of feature names being enabled/disabled, so they have a lot of logic to consider the order in which features were turned on and off, which doesn't scale well as dependency chains get longer.

This patch moves the code handling these dependencies to TargetParser, and changes them to use a Bitset of enabled features. This makes it easy to check which features are enabled, and is converted back to a list of LLVM feature names once all of the command-line options are parsed.

The motivating example for this was the -mcpu=cortex-r82+nofp option. Previously, the code handling the dependency between the fp16 and fp16fml extensions did not consider the nofp modifier, so it added +fullfp16 to the feature list. This should have been disabled by the +nofp modifier, and also the backend did follow the dependency between fullfp16 and fp, resulting in fp being turned back on in the backend.

Most of the dependencies added to AArch64TargetParser.h weren't known about by clang before, I built that list by checking what the backend thinks the dependencies between SubtargetFeatures are.

### CODE REVIEW

**Review by sjoerdmeijer — APPROVED** [review]
This is a very big patch, but it makes a lot of sense. The idea to do it in this way is clearly an improvement. There are quite a number of (new) moving parts involved here, but look reasonable to me so let's give this is a try.  

Thanks for cleaning up the mess (also my mess).

File: clang/test/Driver/aarch64-predres.c [hunk 1]
```diff
@@ -1,8 +1,9 @@
 // RUN: %clang -### --target=aarch64-none-elf -march=armv8a+predres     %s 2>&1 | FileCheck %s
+// RUN: %clang -### --target=aarch64-none-elf -mcpu=cortex-a520         %s 2>&1 | FileCheck %s
 // CHECK: "-target-feature" "+predres"
 // CHECK-NOT: "-target-feature" "-predres"
 
-// RUN: %clang -### --target=aarch64-none-elf -march=armv8.5a+nopredres %s 2>&1 | FileCheck %s --check-prefix=NOPR
+// RUN: %clang -### --target=aarch64-none-elf -mcpu=cortex-a520+nopredres %s 2>&1 | FileCheck %s --check-prefix=NOPR
```
Comment (DavidSpickett) [inline clang/test/Driver/aarch64-predres.c:6]: Did you mean to remove the armv8.5-a line here? As above you added the cortex-a520 instead of replacing the armv8.5 line.

File: clang/test/Driver/aarch64-bf16.c [hunk 2]
```diff
@@ -5,4 +5,4 @@
 // RUN: %clang --target=aarch64 -march=armv8.5a+bf16+nobf16 -### -c %s 2>&1 | FileCheck -check-prefix=GENERICV85A-BF16-NO-BF16 %s
 // GENERICV85A-BF16-NO-BF16: "-target-feature" "-bf16"
 // RUN: %clang --target=aarch64 -march=armv8.5a+bf16+sve -### -c %s 2>&1 | FileCheck -check-prefixes=GENERICV85A-BF16-SVE %s
-// GENERICV85A-BF16-SVE: "-target-feature" "+bf16" "-target-feature" "+sve"
+// GENERICV85A-BF16-SVE: "-target-feature" "+bf16"{{.*}} "-target-feature" "+sve"
```
Comment (DavidSpickett) [inline clang/test/Driver/aarch64-bf16.c:8]: In general, why are these `{{.*}}` now needed in many tests? Is it that we're listing features we didn't before, because before they were expanded in the backend?

Features that are unrelated to, in this case, bf16.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 3]
```diff
@@ -308,6 +312,94 @@ inline constexpr ExtensionInfo Extensions[] = {
 };
 // clang-format on
 
+struct ExtensionSet {
+  // Set of extensions which are currently enabled.
+  ExtensionBitset Enabled;
+  // Set of extensions which have been enabled or disabled at any point. Used
+  // to avoid cluttering the cc1 command-line with lots of unneeded features.
+  ExtensionBitset Touched;
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:320]: Meaning, if extension foo starts as disabled and is never changed or updated, there wouldn't be a "+nofoo" produced?

Which seems correct, just making sure I understand.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 4]
```diff
@@ -308,6 +312,94 @@ inline constexpr ExtensionInfo Extensions[] = {
...
+  // Base architecture version, which we need to know because some feature
+  // dependencies change depending on this.
+  const ArchInfo *BaseArch;
+
+  ExtensionSet() : Enabled(), Touched(), BaseArch(nullptr) {}
+
+  // Enable the given architecture extension, and any other extensions it
+  // depends on.
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:328]: Is it worth saying here whether it updates the base architecture version? I'm guessing that it does not.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 5]
```diff
@@ -308,6 +312,94 @@ inline constexpr ExtensionInfo Extensions[] = {
...
+  void enable(ArchExtKind E);
+  // Disable the given architecture extension, and any other extensions which
+  // depend on it.
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:331]: Is this exhaustive? If `a` depends on `b` and `c`, I would get `a+b+c`. Then if I disable `a`, am I left with `b+c`?

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 6]
```diff
@@ -308,6 +312,94 @@ inline constexpr ExtensionInfo Extensions[] = {
...
+
+  // Add or remove a feature based on a modifier string. The string must be of
+  // the form "<name>" to enable a feature or "no<name>" to disable it. This
+  // will also enable or disable any features as required by the dependencies
+  // between them.
+  bool parseModifier(StringRef Modifier);
+
+  // Convert the set of enabled extension to an LLVM feature list, appending
+  // them to Features.
+  void toLLVMFeatureList(std::vector<StringRef> &Features) const;
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:347]: Is it useful to have this return by ref? Could it be:
```
std::vector<StringRef> toLLVMFeatureList() const;
```
Depends what the majority of call sites do.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 7]
```diff
@@ -308,6 +312,94 @@ inline constexpr ExtensionInfo Extensions[] = {
...
+};
+
+struct ExtensionDependency {
+  ArchExtKind Earlier;
+  ArchExtKind Later;
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:352]: Please add comments to explain the meaning here, I'm not sure what earlier later means.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 8]
```diff
@@ -308,6 +312,94 @@ inline constexpr ExtensionInfo Extensions[] = {
...
+};
+
+// clang-format off
+inline constexpr ExtensionDependency ExtensionDependencies[] = {
+  {AEK_FP, AEK_FP16},
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:357]: Do I understand that this is `{foo, bar}` meaning if you enable `bar` you must add the feature `foo`?

Which seems backwards at first glance but when I get to the use of this maybe it'll be more clear.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 5 above)
Comment (DavidSpickett) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:331]: If this is the case, that's no worse than it ever was.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 7 above)
Comment (DavidSpickett) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:352]: Whatever someone would need if told "hey go make sure Foo implies Bar".

**Review by DavidSpickett — COMMENTED** [review]
Happy to see this change, will be a lot easier to reason about this once it's all in one place.

File: clang/test/Driver/aarch64-bf16.c (hunk 2 above)
Comment (ostannard) [reply clang/test/Driver/aarch64-bf16.c:8]: Yes, we now list every optional feature which the architecture/cpu has by default, most of which aren't relevant to the tests. The order has also changed, so that it matches the source-code order in AArch64TargetParser.h

File: clang/test/Driver/aarch64-predres.c (hunk 1 above)
Comment (ostannard) [reply clang/test/Driver/aarch64-predres.c:6]: Because `armv8.5a` doesn't have the `predres` feature by default, adding `+nopredres` doesn't cause it to appear in the `-###` output. `cortex-a520` does have `predres` by default, so it does.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 3 above)
Comment (ostannard) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:320]: Yes, that's correct.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 4 above)
Comment (ostannard) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:328]: It does not, I'll expand the comment.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 5 above)
Comment (ostannard) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:331]: Yes, that's what happens. That's also the current behaviour on most cases, because it's what the LLVM backend does if passed an inconsistent set of options.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 6 above)
Comment (ostannard) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:347]: The only call-site so far is `getAArch64TargetFeatures` in the driver, which gets passed a reference to an existing `std::vector<StringRef>`, so this way is easier, and avoids copying between vectors.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 7 above)
Comment (ostannard) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:352]: Done. Earlier and later refer to the order in which extensions were added to the architecture, so Later depends on Earlier. I had this as "to" and "from" before, but found that a lot harder to remember which way round they go.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 8 above)
Comment (ostannard) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:357]: Yes, I picked the ordering to match the ordering in which extensions were added to the architecture.

File: clang/test/Driver/aarch64-bf16.c (hunk 2 above)
Comment (DavidSpickett) [reply clang/test/Driver/aarch64-bf16.c:8]: Understood.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 6 above)
Comment (DavidSpickett) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:354]: Cool, leave it as is then.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h (hunk 7 above)
Comment (DavidSpickett) [reply llvm/include/llvm/TargetParser/AArch64TargetParser.h:362]: Yes, if the explanation is "foo was added before bar" then earlier and later is the better choice.

When I was thinking about this problem I always phrased it as "x implies y", which is why it looked backwards to me at first but either is fine as long as it's clearly documented.

**Review by jthackray — APPROVED** [review]
No specific comments, but am in favour of this, as it reduces code complexity and should ensure accuracy of features.

File: llvm/include/llvm/TargetParser/AArch64TargetParser.h [hunk 9]
```diff
@@ -308,6 +312,104 @@ inline constexpr ExtensionInfo Extensions[] = {
...
+  bool parseModifier(StringRef Modifier);
+
+  // Convert the set of enabled extension to an LLVM feature list, appending
+  // them to Features.
+  void toLLVMFeatureList(std::vector<StringRef> &Features) const;
+};
+
+// Represents a dependency between two architecture extensions. If Later is
+// enabled, then Earlier must also be enabled. If Earlier is disabled, then
+// Later must also be disabled.
```
Comment (DavidSpickett) [inline llvm/include/llvm/TargetParser/AArch64TargetParser.h:359]: You should clarify what kind of later this is. "Where feature Earlier was added to the architecture before Later".

As opposed to later on the command line, added later in the compiler, etc.

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Review by tommat01 — COMMENTED** [review]
This looks like a great improvement over the existing system to me.
